### PR TITLE
[extractor/vk] fix video extractor

### PR DESCRIPTION
**Pull request by Mehavoid**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

The other day vk.com dev removed method of getting info for video.
This PR partially fixed issue: https://github.com/yt-dlp/yt-dlp/issues/4097#issuecomment-1159448112

### CODE REVIEW

File: yt_dlp/extractor/vk.py [hunk 1]
```diff
@@ -84,16 +90,19 @@ class VKIE(VKBaseIE):
...
                 'title': 'ProtivoGunz - Хуёвая песня',
                 'uploader': 're:(?:Noize MC|Alexander Ilyashenko).*',
-                'uploader_id': '-77521',
+                'uploader_id': '39545378',
                 'duration': 195,
                 'timestamp': 1329049880,
                 'upload_date': '20120212',
+                'comment_count': int,
+                'like_count': int,
+                'thumbnail': r're:^https?://.*\.jpg$',
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:105]:
```suggestion
                'thumbnail': r're:https?://.+\.jpg$',
```

File: yt_dlp/extractor/vk.py [hunk 2]
```diff
@@ -51,11 +51,17 @@ def _perform_login(self, username, password):
...
-            'https://vk.com/%s.php' % path, video_id,
+            url,
+            video_id,
             data=urlencode_postdata(data), fatal=fatal,
-            headers={'X-Requested-With': 'XMLHttpRequest'})['payload']
+            headers={
+                'Referer': url,
+                'X-Requested-With': 'XMLHttpRequest',
+            },
+        )['payload']
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:64]: revert unnecessary changes.
follow our code style

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by fonclub** [comment]
Work for me. Thank you!

**Comment by dirkf** [comment]
Applying this to yt-dl's almost identical extractor, results are mostly good.

``uploader_id`: '39545378'` is correct in tests 0 and 2 based on `mv_data['authorId']` (`-77521` is a group id). Similarly, the same new file MD5 value is found in these tests.

The test from the yt-dlp extractor with `view_count` fails. `payload_data[1]` is meant to be HTML from which `view_count` can be extracted, but it's always `An error occurred` like this:
```html
<div id="video_box_wrap-77521_162222515" class="video_box_wrap">\n  <div class="video_box_msg">\n    <div class="video_box_msg_background" style="background-image:url(https://sun9-77.userapi.com/c840729/v840729660/45fcc/VCSnkc3l4C4.jpg);"></div>\n    <div class="video_box_msg_text">An error occurred</div\n  </div>\n</div>
```
For embed links like test 2 (unaffected by the current site breakage), the extractor uses a `video_ext.php` page instead, but, although we could fake up an embed URL, using the two parts of the `video_id` and the hash parameter from `player['params'][0]['embed_hash'], the resulting page doesn't have view count data either. That data seems to be out of reach with the existing tactics.

Test 6 (https://vk.com/video-93049196_456239755?list=ln-cBjJ7S4jYYx3ADnmDT) loses `view_count` but needs `timestamp`, `uploader`, `uploader_id`.

In test 9 the video in https://vk.com/video-37468416_456239855 "has been deleted".

In test 11 there's a TODO comment on view_count that's now outdated (since we are using `act=show` and not getting one).

**Comment by finn1199** [comment]
> Work for me. Thank you!

It works nice too!

File: yt_dlp/extractor/vk.py [hunk 3]
```diff
@@ -51,11 +51,17 @@ def _perform_login(self, username, password):
...
-            'https://vk.com/%s.php' % path, video_id,
+            url,
+            video_id,
             data=urlencode_postdata(data), fatal=fatal,
-            headers={'X-Requested-With': 'XMLHttpRequest'})['payload']
+            headers={
+                'Referer': url,
+                'X-Requested-With': 'XMLHttpRequest'
+            }
+        )['payload']
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:64]:
```suggestion
        data['al'] = 1
        code, payload = self._download_json(
            f'https://vk.com/{path}.php', video_id,
            data=urlencode_postdata(data), fatal=fatal,
            headers={
                'Referer': url,
                'X-Requested-With': 'XMLHttpRequest'
            })['payload']
```

File: yt_dlp/extractor/vk.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/vk.py]: No, it is not! You sure you read the guidelines?

**Comment by roy-akash** [comment]
This works for me when i use user and password options.
It does not work for public videos with user and password options.

**Comment by roy-akash** [comment]
It worked intermittently for me , now its not working again .

**Comment by krasnh** [comment]
Работает форк с: 
_--cookies-from-browser_ или
_--cookies FILE_ 

Не работает с:
 _--username_


Example:
`$ ./yt-dlp --cookies=cookies.txt https://vk.com/video-19847999_456242413 
[vk] -19847999_456242413: Downloading JSON metadata
[vk] -19847999_456242413: Downloading m3u8 information
[info] -19847999_456242413: Downloading 1 format(s): hls-2664
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 10
[download] Destination: 22-летний Тьерри Анри терроризирует Канаварро, Несту и всю оборону Италии во время финала Евро-2000. [-19847999_456242413].mp4
[download]  42.7% of ~32.34MiB at    1.85MiB/s ETA 00:09 (frag 4/10)`

**Comment by krasnh** [comment]
The fork works with:
--cookies-from-browser or
--cookies FILE
Doesn't work with:
  --username

Example:
`$ yt-dlp --cookies=cookies.txt https://vk.com/video-19847999_456242413 
[vk] -19847999_456242413: Downloading JSON metadata
[vk] -19847999_456242413: Downloading m3u8 information
[info] -19847999_456242413: Downloading 1 format(s): hls-2664
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 10
[download] Destination: 22-летний Тьерри Анри терроризирует Канаварро, Несту и всю оборону Италии во время финала Евро-2000. [-19847999_456242413].mp4
[download]  42.7% of ~32.34MiB at    1.85MiB/s ETA 00:09 (frag 4/10)`

File: yt_dlp/extractor/vk.py (hunk 2 above)
Comment (Kmartynov) [reply yt_dlp/extractor/vk.py]: Thank you. For me it works.

**Comment by Deathcrash** [comment]
> Работай на меня. Благодарю вас!
почему у меня не работает

**Comment by krasnh** [comment]
> я пробовал вот этот код, что-то не помогло


```
git clone https://github.com/Mehavoid/yt-dlp.git
cd yt-dlp
git checkout fix-vk-video
make yt-dlp

./yt-dlp --cookies-from-browser='firefox:abcdefgh.default' 'https://vk.com/video77496727_456240355'

```

**Comment by fixator10** [comment]
One command ~~to rule them all~~
`pip install --force-reinstall -U git+https://github.com/yt-dlp/yt-dlp/@refs/pull/4128/merge`

**Comment by Deathcrash** [comment]
> Работай на меня. Благодарю вас!

почему у меня не работает я пробовал -cookie and -cookie from browser 

например вот такой код

`from yt_dlp import YoutubeDL

url = 'https://vk.com/video77496727_456240355'
ydl_opts = {"cookies-from-browser": 'firefox:default'}
with YoutubeDL(ydl_opts) as ydl:
    ydl.download(url)
`

**Comment by Deathcrash** [comment]
а как сделать, чтобы работало с python?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/vk.py [hunk 4]
```diff
@@ -30,8 +30,10 @@ class VKBaseIE(InfoExtractor):
     _WAF_URI = 'https://vk.com/429.html?'
 
     def _download_webpage_handle(self, *args, **kwargs):
-        content, urlh = super(VKBaseIE, self)._download_webpage_handle(
-            *args, **kwargs)
+        try:
+            content, urlh = super()._download_webpage_handle(*args, **kwargs)
+        except TypeError:
+            return False
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:35]: too broad! Handle False explictly
```suggestion
        res = super()._download_webpage_handle(*args, **kwargs)
        if not res:
            return res
        content, urlh = res
```

File: yt_dlp/extractor/vk.py [hunk 5]
```diff
@@ -26,6 +28,24 @@
...
+        challenge_url = urlh.geturl()
+        if challenge_url.startswith('https://vk.com/429.html?'):
+            cookie = self._get_cookies(challenge_url).get('hash429')
+            if cookie:
+                hash429 = hashlib.md5(cookie.value.encode('ascii')).hexdigest()
+                resolve_url = update_url_query(challenge_url, {'key': hash429})
+                self._request_webpage(
+                    resolve_url, None,
+                    note='Resolving WAF challenge for VK',
+                    errnote='Failed bypass WAF challenge for VK')
```
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/vk.py:45]: It should follow whatever `fatal` was passed to this function

File: yt_dlp/extractor/vk.py [hunk 6]
```diff
@@ -26,6 +28,24 @@
...
+            cookie = self._get_cookies(challenge_url).get('hash429')
+            if cookie:
+                hash429 = hashlib.md5(cookie.value.encode('ascii')).hexdigest()
+                resolve_url = update_url_query(challenge_url, {'key': hash429})
+                self._request_webpage(
+                    resolve_url, None, fatal=kwargs.get('fatal'),
+                    note='Resolving WAF challenge',
+                    errnote='Failed bypass WAF challenge')
+                return super()._download_webpage_handle(*args, **kwargs)
+        return (content, urlh)
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:47]:
```suggestion
    def _download_webpage_handle(self, url_or_request, video_id, *args, fatal=True, **kwargs):
        response = super()._download_webpage_handle(url_or_request, video_id, *args, fatal=fatal, **kwargs)
        challenge_url, cookie = response[1].geturl() if response else '', None
        if challenge_url.startswith('https://vk.com/429.html?'):
            cookie = self._get_cookies(challenge_url).get('hash429')
        if not cookie:
            return response
            
        hash429 = hashlib.md5(cookie.value.encode('ascii')).hexdigest()
        self._request_webpage(
            update_url_query(challenge_url, {'key': hash429}), video_id, fatal=fatal,
            note='Resolving WAF challenge', errnote='Failed to bypass WAF challenge')
        return super()._download_webpage_handle(url_or_request, video_id, *args, fatal=True, **kwargs)
```

File: yt_dlp/extractor/vk.py [hunk 7]
```diff
@@ -51,11 +71,15 @@ def _perform_login(self, username, password):
                 'Unable to login, incorrect username and/or password', expected=True)
 
     def _download_payload(self, path, video_id, data, fatal=True):
+        endpoint = f'https://vk.com/{path}.php'
         data['al'] = 1
         code, payload = self._download_json(
-            'https://vk.com/%s.php' % path, video_id,
+            endpoint, video_id,
             data=urlencode_postdata(data), fatal=fatal,
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:78]:
```suggestion
        data['al'] = 1
        code, payload = self._download_json(
            f'https://vk.com/{path}.php', video_id, data=urlencode_postdata(data), fatal=fatal,
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pras92** [comment]
> The fork works with: --cookies-from-browser or --cookies FILE Doesn't work with: --username

Doesn't work either way, even with public VK videos.

With --cookies:

```
D:\ytdlp>yt-dlp --cookies D:\ytdlp\cookies.txt https://vk.com/video-139052365_456239297 -v
[debug] Command-line config: ['--cookies', 'D:\\ytdlp\\cookies.txt', 'https://vk.com/video-139052365_456239297', '-v']
[debug] Encodings: locale cp65001, fs utf-8, pref cp65001, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05e] (win32_exe)
[debug] Python 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg git-2020-08-14-a762fd2, ffprobe git-2020-08-14-a762fd2
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] [vk] Extracting URL: https://vk.com/video-139052365_456239297
[vk] -139052365_456239297: Downloading JSON metadata
ERROR: [vk] Unable to download JSON metadata: HTTP Error 404: Not Found (caused by <HTTPError 404: 'Not Found'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 644, in extract
  File "yt_dlp\extractor\vk.py", line 328, in _real_extract
  File "yt_dlp\extractor\vk.py", line 55, in _download_payload
  File "yt_dlp\extractor\common.py", line 1004, in download_content
  File "yt_dlp\extractor\common.py", line 968, in download_handle
  File "yt_dlp\extractor\common.py", line 838, in _download_webpage_handle
  File "yt_dlp\extractor\common.py", line 795, in _request_webpage

  File "yt_dlp\extractor\common.py", line 777, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 3639, in urlopen
  File "urllib\request.py", line 531, in open
  File "urllib\request.py", line 640, in http_response
  File "urllib\request.py", line 569, in error
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 649, in http_error_default
urllib.error.HTTPError: HTTP Error 404: Not Found
```

With --username:

```
D:\ytdlp>yt-dlp --username abc@xyz.com https://vk.com/video-139052365_456239297 -v
[debug] Command-line config: ['--username', 'PRIVATE', 'https://vk.com/video-139052365_456239297', '-v']
Type account password and press [Return]:
[debug] Encodings: locale cp65001, fs utf-8, pref cp65001, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05e] (win32_exe)
[debug] Python 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg git-2020-08-14-a762fd2, ffprobe git-2020-08-14-a762fd2
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[vk] Downloading login page
[vk] Logging in
[debug] [vk] Extracting URL: https://vk.com/video-139052365_456239297
[vk] -139052365_456239297: Downloading JSON metadata
ERROR: [vk] This video is only available for registered users. Use --cookies, --cookies-from-browser, --username and --password, or --netrc (vk) to provide account credentials
  File "yt_dlp\extractor\common.py", line 644, in extract
  File "yt_dlp\extractor\vk.py", line 328, in _real_extract
  File "yt_dlp\extractor\vk.py", line 60, in _download_payload
  File "yt_dlp\extractor\common.py", line 1108, in raise_login_required
  ```

**Comment by krasnh** [comment]
> Doesn't work either way, even with public VK videos.
Changes made to the master. УМВР.
https://github.com/yt-dlp/yt-dlp/commit/59f63c8f0facb71208c8c131935fc4317e96f8b4

**Comment by filkint** [comment]
It doesn't work again.
`yt-dlp -v -F https://vk.com/video-75679763_456275479
[debug] Command-line config: ['-v', '-F', 'https://vk.com/video-75679763_456275479']
[debug] Home config "yt-dlp.conf": []
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05ef6] (zip)
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Python 3.10.4 (CPython 64bit) - Linux-5.15.0-43-generic-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.11.0, brotli-1.0.9, certifi-2020.06.20, pyxattr-0.7.2, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] [vk] Extracting URL: https://vk.com/video-75679763_456275479
[vk] -75679763_456275479: Downloading JSON metadata
ERROR: InfoExtractor._extract_url() takes 2 positional arguments but 3 were given
Traceback (most recent call last):
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/YoutubeDL.py", line 1438, in wrapper
    return func(self, *args, **kwargs)
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/YoutubeDL.py", line 1514, in __extract_info
    ie_result = ie.extract(url)
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/extractor/common.py", line 665, in extract
    ie_result = self._real_extract(url)
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/extractor/vk.py", line 435, in _real_extract
    vimeo_url = VimeoIE._extract_url(url, info_page)
TypeError: InfoExtractor._extract_url() takes 2 positional arguments but 3 were given
`

**Comment by krasnh** [comment]
```
$ yt-dlp  https://vk.com/video-75679763_456275479 
[Cookies] Extracting cookies from firefox
[Cookies] Extracted 445 cookies from firefox
[vk] -75679763_456275479: Downloading JSON metadata
[vk] -75679763_456275479: Downloading m3u8 information
[info] -75679763_456275479: Downloading 1 format(s): hls-5822
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 40
[download] Destination: 'Западный триллер на русский манер' - Иди и смотри сериал 'Ваша честь' (2021) — отец Андрей Ткачёв [-75679763_456275479].mp4
[download]   4.3% of ~301.84MiB at    1.43MiB/s ETA 02:46 (frag 1/40)
```

**Comment by filkint** [comment]
It doesn't work anyway.
`/home/fil/Видео/test/yt-dlp/yt-dlp -v -F https://vk.com/video-75679763_456275419
[debug] Command-line config: ['-v', '-F', 'https://vk.com/video-75679763_456275419']
[debug] Home config "yt-dlp.conf": ['--cookies-from-browser=firefox:/home/fil/snap/firefox/common/.mozilla/firefox/y98rcys1.default-release']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05ef6] (zip)
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Python 3.10.4 (CPython 64bit) - Linux-5.15.0-43-generic-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.11.0, brotli-1.0.9, certifi-2020.06.20, pyxattr-0.7.2, secretstorage-3.3.1, sqlite3-2.6.0
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "/home/fil/snap/firefox/common/.mozilla/firefox/y98rcys1.default-release/cookies.sqlite"
[Cookies] Extracted 772 cookies from firefox
[debug] Proxy map: {}
[debug] [vk] Extracting URL: https://vk.com/video-75679763_456275419
[vk] -75679763_456275419: Downloading JSON metadata
ERROR: InfoExtractor._extract_url() takes 2 positional arguments but 3 were given
Traceback (most recent call last):
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/YoutubeDL.py", line 1440, in wrapper
    return func(self, *args, **kwargs)
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/YoutubeDL.py", line 1516, in __extract_info
    ie_result = ie.extract(url)
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/extractor/common.py", line 665, in extract
    ie_result = self._real_extract(url)
  File "/home/fil/Видео/test/yt-dlp/yt-dlp/yt_dlp/extractor/vk.py", line 435, in _real_extract
    vimeo_url = VimeoIE._extract_url(url, info_page)
TypeError: InfoExtractor._extract_url() takes 2 positional arguments but 3 were given
`

**Comment by krasnh** [comment]
> It doesn't work anyway. 

This is the master? No release yet. Must be compiled with the latest fixes.

```
git clone https://github.com/yt-dlp/yt-dlp.git
cd yt-dlp
make yt-dlp

./yt-dlp https://vk.com/video-75679763_456275419
```

**Comment by krasnh** [comment]
>/home/fil/Видео/test/yt-dlp/yt-dlp -v -F https://vk.com/video-75679763_456275419 [debug] Command-line config: ['-v', '-F', 'https://vk.com/video-75679763_456275419'] [debug] Home config "yt-dlp.conf": ['--cookies-from-browser=firefox:/home/fil/snap/firefox/common/.mozilla/firefox/y98rcys1.default-release'] 

The text should be formatted. ``` at the beginning and at the end.

Example, [quoting-code](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code)

**Comment by basejamping** [comment]
Downloaded master. It still gives an error. Maybe I'm doing something wrong?
ERROR: _extract_url() takes 2 positional arguments but 3 were given

**Comment by pukkandan** [comment]
Open a new issue with relevant verbose log

**Comment by pukkandan** [comment]
Actually, first check if 4080efeb0127150c7a84cdcc0940e0a552fbdf4f fixes the issue
